### DISCUSSION TITLE
How to run pip from inside postUpgradeTasks

### DISCUSSION THREAD

**Discussion by wwuck** [renovatebot/renovate#14769] [Request Help] [answered]
Renovate version: renovate/renovate:32.6.10-slim docker image
Self hosted: Bitbucket Server

I'm trying to follow the suggestion at https://github.com/renovatebot/renovate/issues/2334#issuecomment-1017464091 to run pip-compile inside `postUpgradeTasks`, however when I run it I get a warning in renovate logs `pip-compile: not found`.

I then prepended `pip install pip-tools` to the upgrade task but now see `pip: not found` instead.  Is there a special trick to running `pip` or `pip-compile` inside `postUpgradeTasks`?

`requirements.in`:
```
click==8.0.3
```

`config.js`:
```javascript
module.exports = {
    allowPostUpgradeCommandTemplating: true,
    allowedPostUpgradeCommands: ['^pip-compile', '^cd', '^pip'],
}
```

`renovate.json`:
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "constraints": {
        "python": "3.9"
    },
    "packageRules": [
        {
            "matchManagers": ["pip_requirements"],
            "postUpgradeTasks": {
                "commands": [
                    "cd $(dirname {{{packageFile}}}) && pip-compile --upgrade-package={{{depName}}}=={{{newVersion}}} $(basename {{{packageFile}}})"
                ],
                "fileFilters": ["**/requirements*.txt"]
            }
        }
    ],
    "pip_requirements": {
        "fileMatch": [
            "(^|/)requirements\\.in$"
        ],
        "ignorePaths": [
            "**/requirements.txt"
        ]
    }
}

```

output:
```
 WARN: artifactErrors (repository=RT/test-pip-compile, branch=renovate/click-8.x)
       "artifactErrors": [
         {
           "lockFile": "requirements.in",
           "stderr": "Command failed: cd $(dirname requirements.in) && pip-compile --upgrade-package=click==8.0.4 $(basename requirements.in .txt).in\n/bin/sh: 1: pip-compile: not found\n"
         }
       ]
```

**Comment by viceice** [comment, marked as answer]
You need to install python to slim image or use the full docker image.

**Reply by wwuck** [reply]
Ahh ok thanks! I don’t suppose there is a way to make the post upgrade task run in the python sidecar image?

**Reply by viceice** [reply]
I think there is an open issue about it.

**Reply by wwuck** [reply]
Hmmm ok. A quick GitHub search yielded no results so I can create a new issue if there is no existing one?

**Reply by rarkins** [reply]
I don't recall one. We should make sure there's a feature request to specify which sidecar image to run the tasks in

**Reply by viceice** [reply]
https://github.com/renovatebot/renovate/issues/8111

**Reply by wwuck** [reply]
I switched to the full image and pre-commit is running, but it doesn't appear to follow the python version constraint I set in `renovate.json`.

The diff that the upgrade task adds to `requirements.txt`:
```diff
- # This file is autogenerated by pip-compile with python 3.9
+ # This file is autogenerated by pip-compile with python 3.10
```

Is it possible to get the upgrade task to respect the version constraint configuration?

**Reply by rarkins** [reply]
It's not about "respect". The full image has a fixed version of python installed, so that's what will be used.

**Reply by wwuck** [reply]
Ok, so I should switch back to the slim image and install a specific python version in the image via pyenv or similar?  I guess I could also install it in the full image but I don't need the extra non-python stuff in there.

**Reply by wwuck** [reply]
Is there a better or more recommended way to install a version of python in the image?  How does the full or sidecar image install python?

**Reply by rarkins** [reply]
https://github.com/renovatebot/docker-renovate-full/blob/dd9b2c0660d8cd93b898293dfd9f05388d348281/Dockerfile#L68

**Reply by viceice** [reply]
Install python to the slim image, that currently needs to be done as root.
